### ISSUE TITLE
GetCID: 

### ISSUE THREAD

**Issue by lordshashank** [tellor-io/dataSpecs#143]
### Contact Details

shashanktrivedi1917@gmail.com

### Query Type Name

GetCID

### Description

It would take the link to file and return the CID of the file in location. This would ensure correct decentralized CID calculation for files and getting that data on-chain resolving insolvency cases for both miners and clients in filecoin network resolving major issue.

### Your query parameters

1.   url
       - description: File URL 
       - value type: `string`

### Your response type

- abi_type: ```bytes``` - packed: ```false```

### Dispute Considerations

cid will be converted to bytes and posted on-chain. For dispute consideration anyone can run the functions being used to calculate cid which would be opensourced.

### Suggested Data Sources

A backend would be deployed to calculate cid from any url, and made opensource. An API of deployed instance would be used to fetch data and for dispute consideration anyone can clone and run the functions to check if data provided is correct.

**Comment by tkernell** [comment]
This is a cool idea!

To clarify, let's say I have a cid: `ipfs://bafkreibno34e5vklsc4ubyz6q7py6afhjnyz3eti6azicxdj5qmzxnvwva/`.

Is this how you would want this cid encoded (in solidity)?

```sol
bytes memory cidEncoded = abi.encodePacked("ipfs://bafkreibno34e5vklsc4ubyz6q7py6afhjnyz3eti6azicxdj5qmzxnvwva")
```

Or would you want the `ipfs://` prefix removed?

Also, just so you know, all tellor oracle data is encoded into bytes before being reported. So the above example would be encoded as:

```sol
bytes reportData = abi.encode(cidEncoded)
```

**Comment by lordshashank** [comment]
sorry ```encodePacked``` would not work, changed it to ```encode```.

```ipfs://``` is not needed. So it should be like - 
```bytes memory cidEncoded = abi.encode("bafkreibno34e5vklsc4ubyz6q7py6afhjnyz3eti6azicxdj5qmzxnvwva")```

While using ```encode``` would tellor again encode it?

**Comment by tkernell** [comment]
Yes tellor would encode it again, so if we just define it as a `string`, it'll get put on chain by tellor exactly as you have `cidEncoded` above, without any further encoding.

**Comment by 0xSpuddy** [comment]
@lordshashank Ideally (to decentralize) Tellor reporters could download the file and generate CIDs locally, so we're looking into the challenges with that. I wanted to ask you because you might know best already: any recommended open source repo for using python to generate a CID from files? What are the risks associated with downloading odd files from IPFS?

**Comment by lordshashank** [comment]
I guess the repo you gave would work. I use this go repo [https://github.com/tech-greedy/generate-car](https://github.com/tech-greedy/generate-car) as is more robust and would be deploying my implentation of this only for api returning cid if thats not an issue.

Also about risks, not sure of that never looked on that side of things, being dev have successfully done the flow easily though.

**Comment by 0xSpuddy** [comment]
Commenting again because I want to make sure I understand this flow correctly:
- Someone creates a tip for a unique IPFS Gateway URL that's tellor encoded as queryData / QueryId
- Reporter fetches or generates a CID from the IPFS Gateway URL. (needs to be correct for the data found at the tipped URL)
- Reporter submits the abi encoded CID as the value that goes with the requested queryData / QueryId
- User can then read the reported CID and check if it matches the IPFS stored data correctly
- Reports should be disputed if they the CID calculated by the reporter doesn't match the data found at the Gateway URL

**Comment by lordshashank** [comment]
All correct, just one thing, lets not put constraint on URL to be IPFS Gateway URL, it can be any custom URL from where file can be downloaded.

**Comment by 0xSpuddy** [comment]
PR for adding the FilecoinCID query type https://github.com/tellor-io/dataSpecs/pull/146/files

**Comment by 0xSpuddy** [comment]
closed with pr #146
